### PR TITLE
Cloudformation inllining 

### PR DESCRIPTION
**Pull request by toshke**
# Improvements

1. Docker image updated to run of ruby 2.5

2. `cfpublish` command will show both launch stack url and template url for updating the stack 

3. `MappingParam` changes - if there is no mapping (e.g. `EnvironmentType`) for given mapping parameter, it will bubble to outer component, like rest of the parameters, rather than failing on compilation. 

4. `Cfhighlander::Util::Debug` class available for core developers, to dump cloudformation model at any stage of compilation at the disk. This was useful during this feature development, to narrow down bug hunt to particular compilation phase. 

# New Features

## FnJoin

`FnJoin` added as available intrinsic function in highlander templates. 

## Implements #27 -  Render mode for components

Rendering component resources in resulting cloudformation stack is available in 2 modes. This modes
are controlled using `render` keyword of `Component` DSL statement

`Substack` - creates additional substack for cfhighlander component and points to it using [CloudFormation
Stack](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-stack.html) resource type
This is also default render mode - if no render mode is specified `Substack` will be used

`Inline` - places all defined resources from inner component in outer component cloudformation template. Resources,
Outputs, Conditions, Parameters and Mappings are all inlined - please note that some of the template elements may be renamed in this
process in order to assure unique names. 

There are some limitations when using inline components - Inlined component parameters, having values as outputs from another component (inlined or not)
can't be referenced in component conditions. However, conditions referencing mapping values or parameters passed as mapping values,
are allowed.

**`SIDE EFFECTS`** Side effect of moving from substack based to fully inlined stack may be revealing some of the implicit dependencies within an environment

*Example:* Component A defines Hosted Zone, while component B defines Record Set for given hosted zone. Record set is defined
by referencing Zone Name (rather than ZoneId), meaning there is no explicit dependency between the resources. When both components
are rendered as substack, implicit dependency is created if there is at least one output from component A passed as parameter
to component B. Rendering components inlined removes this implicitly defined dependency, as a consequence stack deletion or creation
may be halted, as record set is being created/deleted before prior the record set. 

**`WARNING`** Be aware of [resource, condition, parameter, output and mapping limits](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cloudformation-limits.html) on a single template
when rendering inner components inlined.

**`EXAMPLE`** All of the VPC resources will be rendered in outer component template, while bastion
will be referenced as substack in example below. 

```ruby
CfhighlanderTemplate do

    Component template:'vpc@1.5.0', name: 'vpc', render: Inline
    Component template:'bastion@1.2.0', name: 'bastion', render: Substack

end
```

### CODE REVIEW

File: spec/data/flatten/src/test.cfhighlander.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
+CfhighlanderTemplate do
+
+  # explicit configuration for vpc component
+  vpc_config = { 'maximum_availability_zones' => 5 }
+
+  hosted_zone = FnJoin('.',[Ref('AWS::Region'), Ref('AWS::AccountId'), 'cfhighlander.info'])
+  vpc_template = 'github:toshke/hl-component-vpc#inline_poc'
```
Comment (aaronwalker) [inline spec/data/flatten/src/test.cfhighlander.rb]: maybe push these to a branch on the upstream repos and they update the test. We can then update the tests once they are merged. Or should be even reference external components in the core tests? maybe create a seperate testsuite?

Comment (toshke) [reply spec/data/flatten/src/test.cfhighlander.rb]: `test.cfhighlander.rb` is actually not used in rspec tests - `c.cfhighlander.rb` is. i've removed this file for commit - it was there as an example.

File: .gitignore [hunk 2]
```diff
@@ -9,7 +9,7 @@
 /test/tmp/
 /test/version_tmp/
 /tmp/
-
+spec/data/flatter/src/out
 # Used by dotenv library to load environment variables.
```
[the remark this answers is not in the document]
Comment (toshke) [reply .gitignore]: done

File: README.md [hunk 3]
```diff
@@ -683,6 +683,46 @@ CfhighlanderTemplate do
 end
 
 ```
+## Render mode for components
+
+Rendering component resources in resulting cloudformation stack is available in 2 modes. This modes
```
[the remark this answers is not in the document]
Comment (toshke) [reply README.md]: updated

**Pull request merged**